packages/loader/container-loader/src/layerCompatState.ts

```diff
@@ -0,0 +1,73 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	checkLayerCompatibility,
+	type ILayerCompatDetails,
+	type ILayerCompatSupportRequirements,
+} from "@fluid-internal/client-utils";
+import type { ICriticalContainerError } from "@fluidframework/container-definitions";
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+
+import { pkgVersion } from "./packageVersion.js";
+
+/**
+ * Loader's compatibility details that is exposed to the Runtime layer.
+ */
+export const LoaderCompatDetails: ILayerCompatDetails = {
+	/**
+	 * The package version of the Loader layer.
+	 */
+	pkgVersion,
+	/**
+	 * The current generation of the Loader layer.
+	 */
+	generation: 1,
+	/**
+	 * The features supported by the Loader layer across the Loader / Runtime boundary.
+	 */
+	supportedFeatures: new Set<string>(),
+};
+
+/**
+ * The requirements that the Runtime layer must meet to be compatible with the Loader.
+ */
+export const RuntimeSupportRequirements: ILayerCompatSupportRequirements = {
+	/**
+	 * Minimum generation that Runtime must be at to be compatible with Loader. Note that 0 is used here for
+	 * Runtime layers before the introduction of the layer compatibility enforcement.
+	 */
+	minSupportedGeneration: 0,
+	/**
+	 * The features that the Runtime must support to be compatible with Loader.
+	 */
+	requiredFeatures: [],
+};
+
+/**
+ * Validates that the Runtime layer is compatible with the Loader.
+ */
+export function validateRuntimeCompatibility(
+	maybeRuntimeCompatDetails: ILayerCompatDetails | undefined,
+	disposeFn: (error?: ICriticalContainerError) => void,
+): void {
+	const layerCheckResult = checkLayerCompatibility(
+		RuntimeSupportRequirements,
+		maybeRuntimeCompatDetails,
+	);
+	if (!layerCheckResult.isCompatible) {
+		const error = new UsageError("Loader is not compatible with Runtime", {
+			version: LoaderCompatDetails.pkgVersion,
+			runtimeVersion: maybeRuntimeCompatDetails?.pkgVersion,
+			generation: LoaderCompatDetails.generation,
+			runtimeGeneration: maybeRuntimeCompatDetails?.generation,
+			minSupportedGeneration: RuntimeSupportRequirements.minSupportedGeneration,
+			isGenerationCompatible: layerCheckResult.isGenerationCompatible,
+			unsupportedFeatures: JSON.stringify(layerCheckResult.unsupportedFeatures),
+		});
+		disposeFn(error);
```

Comment:
I'm annoyed we can't do property bags on error objects like this.  @alexvy86 and I put it up for Eng Inputs consideration but didn't get funding.  I'd say for now do this. A general solution will take more thought, but this will be way easier to consume than individual props.  (`errorDetails` already exists as a column in our 1P telemetry DB)

```suggestion
		const error = new UsageError("Loader is not compatible with Runtime", {
			errorDetails: JSON.stringify({
				version: LoaderCompatDetails.pkgVersion,
				runtimeVersion: maybeRuntimeCompatDetails?.pkgVersion,
				generation: LoaderCompatDetails.generation,
				runtimeGeneration: maybeRuntimeCompatDetails?.generation,
				minSupportedGeneration: RuntimeSupportRequirements.minSupportedGeneration,
				isGenerationCompatible: layerCheckResult.isGenerationCompatible,
				unsupportedFeatures: layerCheckResult.unsupportedFeatures,
			}),
		});
```

Reply:
I was actually thinking if would be nice to put these under `details` like we can do with telemetry properties.

However, these are for users / customers to see. Doesn't having them stringified together make it harder to read?

Reply:
Hmmmmmmm I see your point, there's two use cases - one is us logging, and the other is having the props on the error object.  I think we can optimize for logging... I don't necessarily anticipate people inspecting these error objects.

Future thinking - the error logging infra could pull the telemetry props off and (with a few noted exceptions which would get logged as their own props) put those in a property back like `details` or `errorDetails`.  Then you satisfy both scenarios.

Reply:
I suggest `errorDetails` not `details` to avoid clashing if the log that logs the error also has `details`, they're not properly merged today.

Reply:
I changed it. It should be staightforward to update it if needed.